lib/rubocop/cop/style/hash_transform_methods.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for uses of `_.each_with_object({}) {...}`,
+      # `_.map {...}.to_h`, and `Hash[_.map {...}]` that are actually just
+      # transforming either the keys or the values of a hash, and tries to use
+      # a simpler & faster call to `transform_keys` or `transform_values`
+      # instead.
+      #
+      # This can produce false positives if we are transforming an enumerable
+      # of key-value-like pairs that isn't actually a hash, e.g.:
+      # `[[k1, v1], [k2, v2], ...]`
+      #
+      # This cop should only be enabled on Ruby version 2.5 or newer
+      # (`transform_values` was added in 2.4, and `transform_keys` in 2.5.)
+      #
+      # @example
+      #   # bad
+      #   {a: 1, b: 2}.each_with_object({}) { |(k, v), h| h[k] = v * v }
+      #   {a: 1, b: 2}.map { |k, v| [k.to_s, v] }
+      #
+      #   # good
+      #   {a: 1, b: 2}.transform_values { |v| v * v }
+      #   {a: 1, b: 2}.transform_keys { |k| k.to_s }
+      class HashTransformMethods < Cop
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        EACH_WITH_OBJECT_PATTERN = NodePattern.new(<<~PATTERN)
+          (block
+            ({send csend} !(send _ :each_with_index) :each_with_object (hash))
+            (args
+              (mlhs
+                (arg $_)
+                (arg $_))
+              (arg _memo))
+            ({send csend} (lvar _memo) :[]= $_ $_))
+        PATTERN
+
+        HASH_BRACKETS_MAP_PATTERN = NodePattern.new(<<~PATTERN)
+          (send
+            (const _ :Hash)
+            :[]
+            (block
+              ({send csend} !(send _ :each_with_index) {:map :collect})
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_)))
+        PATTERN
+
+        MAP_TO_H_PATTERN = NodePattern.new(<<~PATTERN)
+          ({send csend}
+            (block
+              ({send csend} !(send _ :each_with_index) {:map :collect})
+              (args
+                (arg $_)
+                (arg $_))
+              (array $_ $_))
+            :to_h)
+        PATTERN
+
+        VALS_MSG = 'Prefer `transform_values` over `each_with_object` or `map`.'
```

Comment:
Btw, here it might be best to dynamically decide on `each_with_object` or `map` (depending on the actual code being analyzed), so the message is more specific.